// @google-cloud/vertexai/build/src/genai/types/common.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListAgentEngineSessionEventsResponse = exports.AppendAgentEngineSessionEventResponse = exports.ListReasoningEnginesSessionsResponse = exports.QueryReasoningEngineResponse = exports.ListReasoningEnginesResponse = exports.Framework = exports.ManagedTopicEnum = exports.AgentServerMode = exports.IdentityType = void 0;
/** The identity type to use for the Reasoning Engine. If not specified, the `service_account` field will be used if set, otherwise the default Vertex AI Reasoning Engine Service Agent in the project will be used. */
var IdentityType;
(function (IdentityType) {
    /**
     * Default value. Use a custom service account if the `service_account` field is set, otherwise use the default Vertex AI Reasoning Engine Service Agent in the project. Same behavior as SERVICE_ACCOUNT.
     */
    IdentityType["IDENTITY_TYPE_UNSPECIFIED"] = "IDENTITY_TYPE_UNSPECIFIED";
    /**
     * Use a custom service account if the `service_account` field is set, otherwise use the default Vertex AI Reasoning Engine Service Agent in the project.
     */
    IdentityType["SERVICE_ACCOUNT"] = "SERVICE_ACCOUNT";
    /**
     * Use Agent Identity. The `service_account` field must not be set.
     */
    IdentityType["AGENT_IDENTITY"] = "AGENT_IDENTITY";
})(IdentityType || (exports.IdentityType = IdentityType = {}));
/** The agent server mode. */
var AgentServerMode;
(function (AgentServerMode) {
    /**
     * Unspecified agent server mode. Do not use.
     */
    AgentServerMode["AGENT_SERVER_MODE_UNSPECIFIED"] = "AGENT_SERVER_MODE_UNSPECIFIED";
    /**
     * Stable agent server mode. This mode has everything stable and well-tested features agent engine offers.
     */
    AgentServerMode["STABLE"] = "STABLE";
    /**
     * Experimental agent server mode. This mode contains experimental features.
     */
    AgentServerMode["EXPERIMENTAL"] = "EXPERIMENTAL";
})(AgentServerMode || (exports.AgentServerMode = AgentServerMode = {}));
/** The managed memory topic. */
var ManagedTopicEnum;
(function (ManagedTopicEnum) {
    /**
     * Unspecified topic. This value should not be used.
     */
    ManagedTopicEnum["MANAGED_TOPIC_ENUM_UNSPECIFIED"] = "MANAGED_TOPIC_ENUM_UNSPECIFIED";
    /**
     * Significant personal information about the User like first names, relationships, hobbies, important dates.
     */
    ManagedTopicEnum["USER_PERSONAL_INFO"] = "USER_PERSONAL_INFO";
    /**
     * Stated or implied likes, dislikes, preferred styles, or patterns.
     */
    ManagedTopicEnum["USER_PREFERENCES"] = "USER_PREFERENCES";
    /**
     * Important milestones or conclusions within the dialogue.
     */
    ManagedTopicEnum["KEY_CONVERSATION_DETAILS"] = "KEY_CONVERSATION_DETAILS";
    /**
     * Information that the user explicitly requested to remember or forget.
     */
    ManagedTopicEnum["EXPLICIT_INSTRUCTIONS"] = "EXPLICIT_INSTRUCTIONS";
})(ManagedTopicEnum || (exports.ManagedTopicEnum = ManagedTopicEnum = {}));
/** Framework used to build the application. */
var Framework;
(function (Framework) {
    /**
     * Unspecified framework.
     */
    Framework["FRAMEWORK_UNSPECIFIED"] = "FRAMEWORK_UNSPECIFIED";
    /**
     * React framework.
     */
    Framework["REACT"] = "REACT";
    /**
     * Angular framework.
     */
    Framework["ANGULAR"] = "ANGULAR";
})(Framework || (exports.Framework = Framework = {}));
/** Response for listing agent engines. */
class ListReasoningEnginesResponse {
}
exports.ListReasoningEnginesResponse = ListReasoningEnginesResponse;
/** The response for querying an agent engine. */
class QueryReasoningEngineResponse {
}
exports.QueryReasoningEngineResponse = QueryReasoningEngineResponse;
/** Response for listing agent engine sessions. */
class ListReasoningEnginesSessionsResponse {
}
exports.ListReasoningEnginesSessionsResponse = ListReasoningEnginesSessionsResponse;
/** Response for appending agent engine session event. */
class AppendAgentEngineSessionEventResponse {
}
exports.AppendAgentEngineSessionEventResponse = AppendAgentEngineSessionEventResponse;
/** Response for listing agent engine session events. */
class ListAgentEngineSessionEventsResponse {
}
exports.ListAgentEngineSessionEventsResponse = ListAgentEngineSessionEventsResponse;
