// @google-cloud/vertexai/build/src/genai/converters/_agentengines_converters.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAgentEngineRequestParametersToVertex = exports.updateAgentEngineConfigToVertex = exports.queryAgentEngineRequestParametersToVertex = exports.queryAgentEngineConfigToVertex = exports.listAgentEngineRequestParametersToVertex = exports.listAgentEngineConfigToVertex = exports.getAgentEngineRequestParametersToVertex = exports.getAgentEngineOperationParametersToVertex = exports.deleteAgentEngineRequestParametersToVertex = exports.createAgentEngineRequestParametersToVertex = exports.createAgentEngineConfigToVertex = void 0;
const common = require("@google/genai/vertex_internal");
function createAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (parentObject !== undefined && fromSpec != null) {
        common.setValueByPath(parentObject, ['spec'], fromSpec);
    }
    const fromContextSpec = common.getValueByPath(fromObject, ['contextSpec']);
    if (parentObject !== undefined && fromContextSpec != null) {
        common.setValueByPath(parentObject, ['contextSpec'], fromContextSpec);
    }
    const fromPscInterfaceConfig = common.getValueByPath(fromObject, [
        'pscInterfaceConfig',
    ]);
    if (parentObject !== undefined && fromPscInterfaceConfig != null) {
        common.setValueByPath(parentObject, ['pscInterfaceConfig'], fromPscInterfaceConfig);
    }
    const fromEncryptionSpec = common.getValueByPath(fromObject, [
        'encryptionSpec',
    ]);
    if (parentObject !== undefined && fromEncryptionSpec != null) {
        common.setValueByPath(parentObject, ['encryptionSpec'], fromEncryptionSpec);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSourcePackages = common.getValueByPath(fromObject, [
        'sourcePackages',
    ]);
    if (parentObject !== undefined && fromSourcePackages != null) {
        common.setValueByPath(parentObject, ['sourcePackages'], fromSourcePackages);
    }
    const fromEntrypointModule = common.getValueByPath(fromObject, [
        'entrypointModule',
    ]);
    if (parentObject !== undefined && fromEntrypointModule != null) {
        common.setValueByPath(parentObject, ['entrypointModule'], fromEntrypointModule);
    }
    const fromEntrypointObject = common.getValueByPath(fromObject, [
        'entrypointObject',
    ]);
    if (parentObject !== undefined && fromEntrypointObject != null) {
        common.setValueByPath(parentObject, ['entrypointObject'], fromEntrypointObject);
    }
    const fromRequirementsFile = common.getValueByPath(fromObject, [
        'requirementsFile',
    ]);
    if (parentObject !== undefined && fromRequirementsFile != null) {
        common.setValueByPath(parentObject, ['requirementsFile'], fromRequirementsFile);
    }
    const fromAgentFramework = common.getValueByPath(fromObject, [
        'agentFramework',
    ]);
    if (parentObject !== undefined && fromAgentFramework != null) {
        common.setValueByPath(parentObject, ['agentFramework'], fromAgentFramework);
    }
    const fromPythonVersion = common.getValueByPath(fromObject, [
        'pythonVersion',
    ]);
    if (parentObject !== undefined && fromPythonVersion != null) {
        common.setValueByPath(parentObject, ['pythonVersion'], fromPythonVersion);
    }
    return toObject;
}
exports.createAgentEngineConfigToVertex = createAgentEngineConfigToVertex;
function createAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        createAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.createAgentEngineRequestParametersToVertex = createAgentEngineRequestParametersToVertex;
function deleteAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromForce = common.getValueByPath(fromObject, ['force']);
    if (fromForce != null) {
        common.setValueByPath(toObject, ['force'], fromForce);
    }
    return toObject;
}
exports.deleteAgentEngineRequestParametersToVertex = deleteAgentEngineRequestParametersToVertex;
function getAgentEngineOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineOperationParametersToVertex = getAgentEngineOperationParametersToVertex;
function getAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.getAgentEngineRequestParametersToVertex = getAgentEngineRequestParametersToVertex;
function listAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromPageSize = common.getValueByPath(fromObject, ['pageSize']);
    if (parentObject !== undefined && fromPageSize != null) {
        common.setValueByPath(parentObject, ['_query', 'pageSize'], fromPageSize);
    }
    const fromPageToken = common.getValueByPath(fromObject, ['pageToken']);
    if (parentObject !== undefined && fromPageToken != null) {
        common.setValueByPath(parentObject, ['_query', 'pageToken'], fromPageToken);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['_query', 'filter'], fromFilter);
    }
    return toObject;
}
exports.listAgentEngineConfigToVertex = listAgentEngineConfigToVertex;
function listAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        listAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.listAgentEngineRequestParametersToVertex = listAgentEngineRequestParametersToVertex;
function queryAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromClassMethod = common.getValueByPath(fromObject, ['classMethod']);
    if (parentObject !== undefined && fromClassMethod != null) {
        common.setValueByPath(parentObject, ['classMethod'], fromClassMethod);
    }
    const fromInput = common.getValueByPath(fromObject, ['input']);
    if (parentObject !== undefined && fromInput != null) {
        common.setValueByPath(parentObject, ['input'], fromInput);
    }
    const fromIncludeAllFields = common.getValueByPath(fromObject, [
        'includeAllFields',
    ]);
    if (fromIncludeAllFields != null) {
        common.setValueByPath(toObject, ['includeAllFields'], fromIncludeAllFields);
    }
    return toObject;
}
exports.queryAgentEngineConfigToVertex = queryAgentEngineConfigToVertex;
function queryAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        queryAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.queryAgentEngineRequestParametersToVertex = queryAgentEngineRequestParametersToVertex;
function updateAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (parentObject !== undefined && fromSpec != null) {
        common.setValueByPath(parentObject, ['spec'], fromSpec);
    }
    const fromContextSpec = common.getValueByPath(fromObject, ['contextSpec']);
    if (parentObject !== undefined && fromContextSpec != null) {
        common.setValueByPath(parentObject, ['contextSpec'], fromContextSpec);
    }
    const fromPscInterfaceConfig = common.getValueByPath(fromObject, [
        'pscInterfaceConfig',
    ]);
    if (parentObject !== undefined && fromPscInterfaceConfig != null) {
        common.setValueByPath(parentObject, ['pscInterfaceConfig'], fromPscInterfaceConfig);
    }
    const fromEncryptionSpec = common.getValueByPath(fromObject, [
        'encryptionSpec',
    ]);
    if (parentObject !== undefined && fromEncryptionSpec != null) {
        common.setValueByPath(parentObject, ['encryptionSpec'], fromEncryptionSpec);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSourcePackages = common.getValueByPath(fromObject, [
        'sourcePackages',
    ]);
    if (parentObject !== undefined && fromSourcePackages != null) {
        common.setValueByPath(parentObject, ['sourcePackages'], fromSourcePackages);
    }
    const fromEntrypointModule = common.getValueByPath(fromObject, [
        'entrypointModule',
    ]);
    if (parentObject !== undefined && fromEntrypointModule != null) {
        common.setValueByPath(parentObject, ['entrypointModule'], fromEntrypointModule);
    }
    const fromEntrypointObject = common.getValueByPath(fromObject, [
        'entrypointObject',
    ]);
    if (parentObject !== undefined && fromEntrypointObject != null) {
        common.setValueByPath(parentObject, ['entrypointObject'], fromEntrypointObject);
    }
    const fromRequirementsFile = common.getValueByPath(fromObject, [
        'requirementsFile',
    ]);
    if (parentObject !== undefined && fromRequirementsFile != null) {
        common.setValueByPath(parentObject, ['requirementsFile'], fromRequirementsFile);
    }
    const fromAgentFramework = common.getValueByPath(fromObject, [
        'agentFramework',
    ]);
    if (parentObject !== undefined && fromAgentFramework != null) {
        common.setValueByPath(parentObject, ['agentFramework'], fromAgentFramework);
    }
    const fromPythonVersion = common.getValueByPath(fromObject, [
        'pythonVersion',
    ]);
    if (parentObject !== undefined && fromPythonVersion != null) {
        common.setValueByPath(parentObject, ['pythonVersion'], fromPythonVersion);
    }
    const fromUpdateMask = common.getValueByPath(fromObject, ['updateMask']);
    if (parentObject !== undefined && fromUpdateMask != null) {
        common.setValueByPath(parentObject, ['_query', 'updateMask'], fromUpdateMask);
    }
    return toObject;
}
exports.updateAgentEngineConfigToVertex = updateAgentEngineConfigToVertex;
function updateAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        updateAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.updateAgentEngineRequestParametersToVertex = updateAgentEngineRequestParametersToVertex;
